### ISSUE TITLE
VQ-VAE commitment loss problem and VQ-loss is missing?

### ISSUE THREAD

**Issue by berniewang8177** [wilson1yan/VideoGPT#39]
Hi, 
I am currently studying the VideoGPT and have some doubts on VQ-VAE losses.

Where is VQ loss?
- In the original [paper](https://arxiv.org/pdf/1711.00937.pdf), there're 3 loss, a reconstruction loss, a VQ loss (bring embedding to encoder_output.detach()). and a commitment loss (bring encoder_output to the embdding.detach()).
- I could only find the commitment loss implemented, but not the VQ loss.

The commitment loss during training is growing after decreasing ...
![image](https://github.com/wilson1yan/VideoGPT/assets/69227993/f81667df-d3fb-4837-b4f9-5018eddda66f)
But I do find my reconstruction loss is decreasing. 

Any information to resolve the situation will be highly appreciated. Is the growing of commitment loss caused by lacking VQ loss?

**Comment by shahrushi2003** [comment]
That is because the codebook is being updated using exponential moving average (EMA), not by the gradient of the codebook loss (see line 176 of vqvae.py). 

It's shown in this [paper](https://arxiv.org/pdf/2305.08842) that EMA-based updation is equivalent to updating the codebook using SGD over codebook loss.
